include/zephyr/drivers/flash.h

```diff
@@ -38,6 +38,26 @@ struct flash_pages_layout {
 };
 #endif /* CONFIG_FLASH_PAGE_LAYOUT */
 
+
+#define FLASH_DEV_INFOWORD(dev) (((struct flash_driver_api *)dev->api)->infoword)
```

Comment:
This should be an inline function, and should copy the infoword to be compatible with userspace
```
__syscall int flash_get_infoword(const struct device *dev, struct *infoword);
...
```

Reply:
The Kconfig options are provided to optimize how decisions in code are taken. For example if you only have a device that requires erase, then you will only have a device that selects `FLASH_DEVICE_REQUIRES_ERASE`.
In such case the:
```
FLASH_ERASE_REQUIRED(dev)
```
evaluates always to true and this optimizes code, as now the code know that there is not device that does not require erase and will not check parameters.
In most cases users will have only one kind of device, in which case run-time checking of these is waste of clock and flash.

Reply:
Considering we need to read the write_block_size and often erase_value anyway, if we have one structure with device info, the potential ROM saving from this compile time variable is not worth it IMO

Reply:
Once you find out that you only address devices that do not require erase, you are not going to need above parameters.
I have also planned to move these into infoword, as I did in ZSAI, but a this point I have decided to reduce approach to have some success at least in making this go further.

Reply:
Syscall takes a lot of time, which makes picking infoword every time you need to check what you are working with, inconvenient. Also you are able to figure a lot at compile time, which means that you can already save clock and flash before run-time.

Reply:
Then I don't understand why we are using the flash API, if no_erase means that we don't need to erase, we don't need to deal with write alignment, and we are not dealing with pages, why not use the retention API?

Reply:
> Then I don't understand why we are using the flash API, if no_erase means that we don't need to erase, we don't need to deal with write alignment, and we are not dealing with pages, why not use the retention API?

Ah sorry, the write-block-size remains.

Reply:
> Syscall takes a lot of time, which makes picking infoword every time you need to check what you are working with, inconvenient. Also you are able to figure a lot at compile time, which means that you can already save clock and flash before run-time.

Accessing device info should always be done using a syscall. This is a very good reason not to add this information at the device level but instead at a higher level based upon the device type.